Add an `external_docker_image` option

florentsfl opened the pull request:
This a porting to bats of the PR #191 

Basically this PR clarifies the role of `custom_img_name` by making it clear that we can specify an external docker image. It was implemented to mimic the behavior of when we use the Dockerfile, that is to say print a warning to re-init when the image is outdated. 

Another solution would be to do as what was done before pulling in all cases when doing cqfd run

felixlechat1 committed:
cqfd: allow cqfd to run without a dockerfile

When using a prebuilt distant docker image, cqfd run will now work when
used with no local dockerfile.

The variable `custom_img_name` had a confusing role, because it was used
both for tagging an image and for pulling a distant image. This commit
makes a role clearer by introducing `external_docker_image` which is
clearly used for pulling a distant image.

Having a way to give an alternate name for the tag of the image
generated by cqfd init has no clear use case

gportay committed:
cqfd: pull external_docker_image on init

cqfd requires the image to be built using the command init.

But, cqfd pulls the custom image set in custom_img_name on command run.

This modifies and renames the docker_build function to either pull the
custom image set, or build it using the Dockerfile.

gportay committed:
tests: 05-cqfd_init_custom_image: rewrite pretest code

The code testing if the custom image does not exist before running the
test creating it **IS NOT** an actual test.

It is a safeguard to ensure the custom image is not already present and
to make sure the test actually creates it.

Note: The custom image remains created if the tests are interrupted
before the created image is cleaned up. Nevertheless, the risks are
mitigated by the use of the double $RANDOM suffix in the custom image
name, which "guarantee" to run multiple batches of test at the same
time.

This rewrites the code and transforms it as a none test case code, as in
05-cqfd_run_c_flavor[1].

[1]: https://github.com/savoirfairelinux/cqfd/blob/master/tests/05-cqfd_run_c_flavor#L15-L20

florentsfl committed:
cqfd: group common arguments

Instead of setting them at various places

deribaucourt commented on README.md:
```diff
@@ -219,15 +219,17 @@ required to build the image.  When using this option, a
 `.dockerignore` file can be useful to limit what gets sent to the
 Docker daemon.
 
-`custom_img_name` (optional): allows specifying a custom Docker image name
```
This is an API breaking change, it should be documented in a Changelog file.
From the previous documentation, it looked like this variable could serve two roles. Your new proposition only supports one of them?
 1. Pulling an external image (ported)
 2. Building the image locally, but with a specific name (missing)
 
I don't think the missing use case is very important, but someone might?
Alternative: keep the previous variable for the missing use case, and split the "pull" feature into the new variable?

florentsfl replied:
I was also thinking that the missing use case is not relevant, so I removed it. But this is a breaking change indeed.

I will try to do as you suggest: keep the old variable for compatibility, and use the new one for clarity

deribaucourt commented on cqfd:
```diff
@@ -728,21 +739,37 @@ load_config() {
-		local dockerfile_hash
+	cqfd_image="$project_docker_run_image"
+	if [ ! "$cqfd_image" ]; then
+		local Un
```
Why did you need to replace the format? The new syntax is harder to read.

florentsfl replied:
yes, this was added to allow to use formats in the `custom_img_name` etc. but since I removed this feature from this PR it is not necessary anymore. I will clean it

deribaucourt commented on tests/05-cqfd_init_custom_image.bats:
```diff
@@ -1,52 +0,0 @@
-    assert_line --regexp "Ubuntu 16.04(.[[:digit:]]+)? LTS"
-}
-
-@test "'cqfd init' creates custom_img_name" {
```
This is the removed feature test case, if we change our mind and restore it, we'll have to restore that test.

florentsfl committed:
cqfd: check if local image is outdated

Just as we are checking if the Dockerfile has changed when we use a
Dockerfile, we should check if the specified image has been updated
online before using it

deribaucourt commented on README.md:
````diff
@@ -223,6 +223,18 @@ Docker daemon.
+it can be set like:
+
+```
+external_docker_image=ubuntu:16.04
````
Let's recommend 26.04 to be more up-to-date :p

florentsfl replied:
done !

deribaucourt commented on tests/05-cqfd_external_docker_image.bats:
```diff
@@ -0,0 +1,52 @@
+    mv -f .cqfd/docker/Dockerfile .cqfd/docker/Dockerfile.old
+    cp -f .cqfdrc .cqfdrc.old
+    cp -f cqfdrc-external_docker_image .cqfdrc
+    export external_docker_image="ubuntu:16.04"
```
Test with something more recent, at least 24.04.

florentsfl replied:
Yes, the goal was to test with an image different than the one which is already included in the base dockerfile, which is 24.04. So I changed it to ubuntu 22.04

deribaucourt commented on tests/05-cqfd_external_docker_image.bats:
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bats
+
+setup_file() {
+    mv -f .cqfd/docker/Dockerfile .cqfd/docker/Dockerfile.old
```
Just like we did in another recent PR, I suggest to use a temporary, separate .cqfd directory to avoid corrupting the repository if the test script is interrupted.

florentsfl replied:
If the test script is interrupted, then is corrupts a temporary directory in /tmp/bats… which is in all cases destroyed after the tests fail. 

So it could eventually corrupts a directory which will in all cases be destroyed right after it is corrupted so there is very little chance that it causes a mess.

And tests are much more readable if we don’t use -d and -f options of cqfd at each call

deribaucourt commented on tests/05-cqfd_external_docker_image.bats:
```diff
@@ -0,0 +1,52 @@
+}
+
+@test "'cqfd' does not pull image" {
+    run cqfd
```
The tests would fail if run out of order in the current implementation. But I think it's not something we are aiming at.

florentsfl replied:
yes we would enforce running these tests in sequence if we enable global parallelization

deribaucourt commented on tests/05-cqfd_init_custom_image.bats:
```diff
@@ -1,43 +1,57 @@
-org="cqfd"
-name="custom-image"
-custom_img_name="$custom_image_1"
+@test "cqfd pulls image custom_img_name from registry" {
```
I though that this behavior only concerned the new "external_docker_image" variable? custom_img_name would be for cqfd init built images.

florentsfl replied:
if we want to keep the current behavior of custom_img_name, then it is also used to pull from a distant registry.

external_docker_image is a replacement for the feature of custome_img_name which pulls from distant registry, with cleaner implementation

deribaucourt replied:
I am not sure there's an interest in keeping the pull mode of custom_img_name since there's a cleaner dedicated implementation for that now.
The naming also seems more aligned with just building under a custom name.

deribaucourt reviewed (changes requested):
I'm OK with the implementation. Some remaining minor remarks, mostly on testing.
